packages/plugin-compat/extra/fsevents/common.patch

```diff
@@ -2,45 +2,108 @@ diff --git a/vfs.js b/vfs.js
-+  const match = p.match(VIRTUAL_REGEXP);
-+  if (!match)
-+    return p;
-+
-+  const target = path.dirname(match[1]);
-+  if (!match[3] || !match[4])
-+    return target;
++  const pnpapi = getPnpApi();
++  if (pnpapi) {
++    const resolved = pnpapi.resolveVirtual(p);
++    // pnp resolveVirtual will return null for non-virtual path.
++    return resolved == null ? p : resolved;
++  }
++  return p;
++}
 +
-+  const isnum = NUMBER_REGEXP.test(match[4]);
-+  if (!isnum)
-+    return p;
++function findVirtualPaths(root) {
++  const paths = [];
++  const pnpapi = getPnpApi();
++  if (pnpapi) {
++    for (const locator of pnpapi.getDependencyTreeRoots()) {
++      const pkg = pnpapi.getPackageInformation(locator);
++      for (const [name, referencish] of pkg.packageDependencies) {
++        if (referencish === null) continue;
++        if (referencish.indexOf(`virtual:`) === 0) {
++          const virtualLocator = pnpapi.getLocator(name, referencish);
++          const virtualPkg = pnpapi.getPackageInformation(virtualLocator);
++          if (virtualPkg && virtualPkg.packageLocation.indexOf(root) === 0) {
++            // virtual path fall under root
++            paths.push(virtualPkg.packageLocation);
++          }
++        }
++      }
++    }
++  }
++  return paths;
++}
 +
-+  const depth = Number(match[4]);
-+  const backstep = `../`.repeat(depth);
-+  const subpath = (match[5] || `.`);
++class VFS {
++  constructor(p, Native) {
++    this.root = path.resolve(p);
++    this.native = Native;
++    this.watchers = [];
++  }
 +
-+  return resolveVirtual(path.join(target, backstep, subpath));
-+}
++  transpose(rawPath, resolvedPath, p) {
++    const transposePath = rawPath + p.substr(resolvedPath.length);
++    return transposePath;
++  }
 +
-+module.exports = class FsePnp {
-+  constructor(p) {
-+    this.normalizedPath = path.resolve(p);
-+    this.resolvedPath = resolveVirtual(this.normalizedPath);
++  /**
++   * build raw and resolved path mapping
++   * @param {*} root
++   */
++  buildPathMap() {
++    const pathMap = new Map();
++    this.resolvedRoot = resolveVirtual(this.root);
++    pathMap.set(this.resolvedRoot, this.root);
++    if (!path.extname(this.root)) {
++      // find all direct virtual paths for given root.
++      const virtualPaths = findVirtualPaths(this.root);
++      virtualPaths.forEach((virtualPath) => {
++        const resolvedVirtual = resolveVirtual(virtualPath);
++        if (resolvedVirtual.indexOf(this.resolvedRoot) < 0) {
++          pathMap.set(resolvedVirtual, virtualPath);
++        }
++      });
++    }
++    return pathMap;
 +  }
 +
-+  transpose(p) {
-+    return this.normalizedPath + p.substr(this.resolvedPath.length);
++  watch(handler) {
++    const pathMap = this.buildPathMap();
++    pathMap.forEach((virtualPath, resolvedPath) => {
++      const watcher = this.native.start(resolvedPath, (p, ...args) => {
++        return handler(this.transpose(virtualPath, resolvedPath, p), ...args);
++      });
```

Comment:
I don't understand why it's necessary to spawn multiple watchers - the case you're trying to address is when calling `watch` on, for example, the root directory of the repository, right? If so, how would multiple watchers help?